### PR TITLE
AppGroups Summarys Impl

### PR DESCRIPTION
**Pull request by RonGonz-illumio**
Implement App Groups Summary

- [X] I've read and followed the [contribution guidelines](CONTRIBUTING.md) for this PR

### CODE REVIEW

File: .github/workflows/run-tests.yml [hunk 1]
```diff
@@ -27,6 +27,10 @@ jobs:
         exclude:
           - python-version: "3.6"
             os: ubuntu-latest
+          - python-version: "3.6"
```
Comment (dsommerville-illumio) [inline .github/workflows/run-tests.yml:30]: Please add a comment here re. the lack of `arm64` support for 3.6 and 3.7.

File: illumio/__init__.py [hunk 2]
```diff
@@ -20,12 +20,13 @@
...
+from .appgroups import *
 
 from types import ModuleType
 
 # avoid name conflicts with package modules when using
 # `from illumio import *` by excluding them here
 __all__ = [
     export for export, o in globals().items()
-        if not (export.startswith('_') or isinstance(o, ModuleType))
+    if not (export.startswith('_') or isinstance(o, ModuleType))
```
Comment (dsommerville-illumio) [inline illumio/__init__.py:31]: Nit: Is this from an automated formatter? Please leave these as-is.

File: illumio/pce.py [hunk 3]
```diff
@@ -503,7 +503,10 @@ def get(self, policy_version: str = DRAFT, parent: Union[str, Reference, dict] =
             """
             endpoint = self._build_endpoint(policy_version, parent)
             response = self.pce.get(endpoint, **{**kwargs, **{'include_org': False}})
-            return [self.object_cls.from_json(o) for o in response.json()]
+            if islist(type(response.json())):
+                return [self.object_cls.from_json(o) for o in response.json()]
+            elif type(response.json()) is dict:
+                return self.object_cls.from_json(response.json())
```
Comment (dsommerville-illumio) [inline illumio/pce.py:509]: Please add another return after the conditional check here - if the value is not a list or a dict we'd currently return `None`. We should instead probably just return `response.json()`.

File: illumio/util/constants.py [hunk 4]
```diff
@@ -46,8 +46,8 @@
 #: See https://www.iana.org/assignments/icmp-parameters/icmp-parameters.xhtml#icmp-parameters-types
 ICMP_TYPE_MAX = 255
 
-FQDN_REGEX = re.compile('(?=^.{4,253}$)(^((?!-)[a-zA-Z0-9-]{1,63}(?<!-)\.)+[a-zA-Z]{2,63}$)')
-HREF_REGEX = re.compile('^\/orgs\/\d+\/(?:sec_policy\/(?:active|draft)\/)?(?P<type>[a-zA-Z_]+)\/(?P<uid>[a-zA-Z0-9-]+)$')
+FQDN_REGEX = re.compile(r'(?=^.{4,253}$)(^((?!-)[a-zA-Z0-9-]{1,63}(?<!-)\.)+[a-zA-Z]{2,63}$)')
```
Comment (dsommerville-illumio) [inline illumio/util/constants.py:49]: What does this fix? These regexes were working before.

File: illumio/util/functions.py [hunk 5]
```diff
@@ -63,6 +63,7 @@ def deprecated(deprecated_in, message=None):
     Deprecation decorator, adapted from https://stackoverflow.com/a/30253848
     Will emit a warning when the decorated function is called.
     """
+
```
Comment (dsommerville-illumio) [inline illumio/util/functions.py:66]: Nit: Please turn off the auto-formatter. These newlines are not necessary.

File: tests/unit/mocks.py [hunk 6]
```diff
@@ -56,8 +59,8 @@ class PCEObjectMock(object):
     """
     Base class for PCE object mocks
     """
-    base_pattern = '^/api/v2(?:/orgs/\d+)?(?:/sec_policy/(?:draft|active))?/([a-zA-Z_\-]+)'
-    href_pattern = '^/api/v2((?:/orgs/\d+)?(?:/sec_policy/(?:draft|active))?(?:/[a-zA-Z_\-]+/[a-zA-Z0-9\-]+)+)$'
+    base_pattern = r'^/api/v2(?:/orgs/\d+)?(?:/sec_policy/(?:draft|active))?/([a-zA-Z_\-]+)'
```
Comment (dsommerville-illumio) [inline tests/unit/mocks.py:62]: Again, not sure why these have been changed as they were working fine before.

**Review by dsommerville-illumio — COMMENTED** [review]
Thanks! Looks good, mostly just a couple of nits below.

File: illumio/util/constants.py (hunk 4 above)
Comment (RonGonz-illumio) [reply illumio/util/constants.py:49]: All of the regexes had the following types of issues:

* PEP 8: W605 invalid escape sequence '\+'
* Redundant character escape '\+' in RegExp

Explanation:
In a regular expression, \w is a special sequence that matches any word character (alphanumeric characters and underscores). When using a regular string literal, the backslash \ is treated as an escape character, so \w is interpreted as an invalid escape sequence.

By using a raw string literal with r, the backslash is treated as a literal character, and \w is correctly interpreted as the special sequence for matching word characters.
So, the corrected regular expression pattern r'^\w+://' will match strings that start with one or more word characters followed by ://.  

For example:
http:// will match
ftp:// will match
file:// will match
:// will not match (because it doesn't start with word characters)

Using raw string literals is a good practice when working with regular expressions in Python to avoid issues with escape sequences.

File: illumio/__init__.py (hunk 2 above)
Comment (RonGonz-illumio) [reply illumio/__init__.py:31]: I can do that but it won't be PEP8 compliant.

File: tests/unit/mocks.py (hunk 6 above)
Comment (RonGonz-illumio) [reply tests/unit/mocks.py:62]: Here we have redundant escape characters as well as PEP8 W605 Invalid escape sequences, converting to raw string literal helps avoid these issues.

File: illumio/__init__.py (hunk 2 above)
Comment (RonGonz-illumio) [reply illumio/__init__.py:31]: image

Comment (dsommerville-illumio) [reply illumio/__init__.py]: PEP8 isn't a compliance framework, it's a style guide. The whole point is to improve readability through consistently applied rules. In this case, the unaligned version is more readable because it indicates that the condition is a filter on the `for`.

Comment (RonGonz-illumio) [reply illumio/__init__.py]: already resolved in latest commit - i added the indent back.

File: illumio/appgroups/appgroupsummary.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+from dataclasses import dataclass
+from datetime import datetime
+from enum import Enum
+from typing import List, Dict, Optional
+
+from illumio.util import pce_api, JsonObject
+
+
+class visibility_level_key_types(Enum):
```
Comment (dsommerville-illumio) [inline illumio/appgroups/appgroupsummary.py:21]: Nit: These class names should use `CapitalCase`.

Comment (dsommerville-illumio) [reply illumio/appgroups/appgroupsummary.py:21]: Also, please move them to `constants.py` and extend `IllumioEnumMeta` as with other enums.

File: illumio/util/functions.py [hunk 8]
```diff
@@ -178,63 +175,46 @@ def convert_protocol(protocol: str) -> int:
...
 
     Raises:
         IllumioIntegerValidationException: if an invalid value is provided.
     """
     try:
-        valid = minimum <= int(val) <= maximum
-    except:  # catch the ValueError for invalid values
-        valid = False
-    if not valid:
+        if not minimum <= int(val) <= maximum:
```
Comment (dsommerville-illumio) [inline illumio/util/functions.py:189]: Why was this changed?

File: illumio/util/functions.py [hunk 9]
```diff
@@ -178,63 +175,46 @@ def convert_protocol(protocol: str) -> int:
...
+            raise IllumioIntegerValidationException(val, minimum, maximum)
+    except ValueError:
         raise IllumioIntegerValidationException(val, minimum, maximum)
 
+
 if hasattr(typing, 'get_origin'):
-    # python 3.8+ - introduces the get_origin function
```
Comment (dsommerville-illumio) [inline illumio/util/functions.py:200]: Why remove this comment?

File: illumio/util/functions.py [hunk 10]
```diff
@@ -178,63 +175,46 @@ def convert_protocol(protocol: str) -> int:
...
+    # Python 3.8+
     def isunion(type_):
         return typing.get_origin(type_) is typing.Union
 
     def islist(type_):
-        if type_ is list:
-            return True
-        return typing.get_origin(type_) is list
+        return type_ is list or typing.get_origin(type_) is list
```
Comment (dsommerville-illumio) [inline illumio/util/functions.py:201]: Please don't change these functions.

File: illumio/util/functions.py [hunk 11]
```diff
@@ -178,63 +175,46 @@ def convert_protocol(protocol: str) -> int:
...
 
 elif hasattr(typing, '_GenericAlias'):
-    # python 3.7 - changes meta types to be based off the
```
Comment (dsommerville-illumio) [inline illumio/util/functions.py:210]: Please revert all the changes you've made here.

File: illumio/util/functions.py (hunk 9 above)
Comment (RonGonz-illumio) [reply illumio/util/functions.py:200]: readded

File: illumio/util/functions.py (hunk 8 above)
Comment (RonGonz-illumio) [reply illumio/util/functions.py:189]: seems like it could be optimized - is the optimized version not better?

**Review by dsommerville-illumio — APPROVED** [review]
LGTM, thanks!
